syntax = "proto3";

package pb;
option go_package = "../grpc";

message Base {
    int64 RetCode = 1;
    string RetMsg = 2;
    map<string, string> Extra = 3;
}

message InboundBandwidthUsage {
    repeated double inbound_bandwidth_usage = 1;
}

message OutboundBandwidthUsage {
    repeated double outbound_bandwidth_usage = 1;
}

message ComputationResourceUsage {
    repeated double computation_resource_usage = 1;
}

message Location {
    double latitude = 1;
    double longitude = 2;
}

message UserInfo {
    Location location = 1;
    string channel_id = 2;
    int64 version = 3;
    string user_id = 4;
}

message V2Number {
    map<string, int64> number = 1;
}

message H2V {
    map<string, V2Number> h2v = 1;
}

message ViewerConnection {
    map<string, H2V> viewer_connection_table = 1; // e->h->v->number
}

message QoEPreference {
    float alpha1 = 1;
    float alpha2 = 2;
    float alpha3 = 3;
}

message State {
    InboundBandwidthUsage inbound_bandwidth_usage = 1;
    OutboundBandwidthUsage outbound_bandwidth_usage = 2;
    ComputationResourceUsage computation_resource_usage = 3;
    QoEPreference qoe_preference = 4;
    UserInfo user_info = 5;
    ViewerConnection viewer_connection = 6;
}

message Action {
    string viewer_id = 1;
    string channel_id = 2;
    int64 version = 3;
    repeated double possibility = 4;
    int64 action = 5;
    QoEPreference qoe_preference = 6;
}

message Feedback {
    double reward = 1;
    double accuracy = 2;
}

message BandWidthInfo {
    double inbound_bandwidth_usage = 1;
    double outbound_bandwidth_usage = 2;
    double inbound_bandwidth_limit = 3;
    double outbound_bandwidth_limit = 4;
}

message Device {
    int32 id = 1;
    string name = 2;
    int32 cpu_core = 3;
    Location location = 4;
    BandWidthInfo band_width_info = 5;
    double latency_to_upper = 6;
    double computation_usage = 7;
}

message SystemInfo {
    repeated Device edges = 1;
    repeated Device cdn = 2;
}

message Solve {
    UserInfo user_info = 1;
    string device_name = 2;
}

message TaskManagerInfo {
    int64 time = 1;
    repeated UserInfo user_info = 2;
    repeated Solve solved = 3;
}

message BackgroundInfo {
    int64 time = 1;
    int64 max_time = 2;
    Location location = 3;
}